presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java

```diff
@@ -0,0 +1,398 @@
+        }));
+    }
+
+    public synchronized Map<String, MemoryInfo> getWorkerMemoryInfo()
+    {
+        return nodeStatuses.entrySet().stream().collect(toImmutableMap(e -> {
+            String nodeIdentifier = e.getValue().getNodeStatus().getNodeId();
+            String nodeHost = URI.create(e.getValue().getNodeStatus().getExternalAddress()).getHost();
+            return nodeIdentifier + " [" + nodeHost + "]";
+        }, e -> e.getValue().getNodeStatus().getMemoryInfo()));
+    }
+
+    private static class CoordinatorQueriesState
+    {
+        private final String nodeId;
+        private final int maxCompletedQueries;
+        private final long queryExpirationTimeoutMillis;
+        private final long completedQueryExpirationTimeoutMillis;
+
+        @GuardedBy("this")
+        private final Map<QueryId, Query> activeQueries = new HashMap<>();
+        @GuardedBy("this")
+        private final Map<QueryId, Query> completedQueries = new LinkedHashMap<>();
+
+        public CoordinatorQueriesState(String nodeId, int maxCompletedQueries, long queryExpirationTimeoutMillis, long completedQueryExpirationTimeoutMillis)
+        {
+            this.nodeId = requireNonNull(nodeId, "nodeId is null");
+            checkArgument(maxCompletedQueries > 0);
+            checkArgument(queryExpirationTimeoutMillis > 0);
+            checkArgument(completedQueryExpirationTimeoutMillis > 0);
+            this.maxCompletedQueries = maxCompletedQueries;
+            this.queryExpirationTimeoutMillis = queryExpirationTimeoutMillis;
+            this.completedQueryExpirationTimeoutMillis = completedQueryExpirationTimeoutMillis;
+        }
+
+        public synchronized CoordinatorQueriesState addOrUpdateQuery(BasicQueryInfo basicQueryInfo)
+        {
+            requireNonNull(basicQueryInfo, "basicQueryInfo is null");
+            QueryId queryId = basicQueryInfo.getQueryId();
+            Query query = activeQueries.get(queryId);
+            if (query == null) {
+                query = new Query(basicQueryInfo);
+                activeQueries.put(queryId, query);
+            }
+            else {
+                query = query.updateQueryInfo(basicQueryInfo);
+            }
+            if (isQueryCompleted(query)) {
+                completedQueries.put(query.getQueryId(), query);
+                activeQueries.remove(query.getQueryId());
+            }
+            return this;
+        }
+
+        public synchronized void purgeExpiredQueries()
+        {
+            long currentTimeMillis = System.currentTimeMillis();
+
+            Iterator<Query> queryIterator = activeQueries.values().iterator();
+            while (queryIterator.hasNext()) {
+                Query query = queryIterator.next();
+                if (isQueryExpired(query, currentTimeMillis, queryExpirationTimeoutMillis)) {
+                    completedQueries.put(query.getQueryId(), query);
+                    queryIterator.remove();
+                }
+            }
```

Comment:
If we use concurrent data structure, we likely can still make this part work: https://www.javaer101.com/en/article/499093.html